core/arch/arm/plat-stm32mp1/main.c

```diff
@@ -298,16 +298,25 @@ vaddr_t stm32mp_bkpreg(unsigned int idx)
 
 vaddr_t stm32_get_gpio_bank_base(unsigned int bank)
 {
-	static struct io_pa_va gpios_nsec_base = { .pa = GPIOS_NSEC_BASE };
-	static struct io_pa_va gpioz_base = { .pa = GPIOZ_BASE };
+	static struct io_pa_va x_base = { .pa = GPIOS_NSEC_BASE };
+	static struct io_pa_va z_base = { .pa = GPIOZ_BASE };
 
-	if (bank == GPIO_BANK_Z)
-		return io_pa_or_va(&gpioz_base);
+	/* Get non-secure mapping address for GPIOZ */
+	if (bank == GPIO_BANK_Z) {
+		if (!cpu_mmu_enabled())
+			return (vaddr_t)z_base.pa;
+
+		if (!z_base.va)
+			z_base.va = (vaddr_t)phys_to_virt(z_base.pa,
+							  MEM_AREA_IO_NSEC);
+
+		return z_base.va;
+	}
 
 	COMPILE_TIME_ASSERT(GPIO_BANK_A == 0);
 	assert(bank <= GPIO_BANK_K);
 
-	return io_pa_or_va(&gpios_nsec_base) + (bank * GPIO_BANK_OFFSET);
+	return io_pa_or_va(&x_base) + (bank * GPIO_BANK_OFFSET);
```

Comment:
I wonder if `io_pa_or_va()` should rather be replaced with `io_pa_or_va_sec()` and `io_pa_or_va_nsec()`. I guess that for all invocations you know if you're targeting secure or non-secure IO-memory.

I'm not asking that we should change something here, it's just food for thought.

Reply:
I agree that a generic `io_pa_or_va_{secure|nsec}()` would be nice.
@jforissier, should we keep the legacy `io_pa_or_va()`?

Reply:
> @jforissier, should we keep the legacy `io_pa_or_va()`?

Not necessarily, if it's not needed we could as well remove it. Even if some use case comes later, we can always add it back.

Reply:
I have created P-R https://github.com/OP-TEE/optee_os/pull/3020 for this.

Note that most optee_os drivers using `io_pa_or_va()` are UART console drivers. They can be used on secure or non-secure peripherals. Therefore I did not remove `io_pa_or_va()` as still used from these drivers.